File: patterns/cache-aside/examples/cache_aside.js
/**
 * Cache-Aside Pattern - Node.js Implementation
 * =============================================
 */

const redis = require('redis');
const { Pool } = require('pg');

class CacheAsideRepository {
  constructor(redisClient, pgPool) {
    this.cache = redisClient;
    this.db = pgPool;
    this.defaultTTL = 3600; // 1 hour
    this.stats = {
      hits: 0,
      misses: 0,
      writes: 0
    };
  }

  /**
   * Get user with cache-aside pattern
   */
  async getUser(userId) {
    const cacheKey = `user:profile:${userId}`;

    try {
      // 1. Check cache first
      const cached = await this.cache.get(cacheKey);
      
      if (cached) {
        this.stats.hits++;
        console.log(`✓ Cache HIT: ${cacheKey}`);
        
        if (cached === 'NOT_FOUND') {
          return null;
        }
        
        return JSON.parse(cached);
      }

      // 2. Cache miss - query database
      this.stats.misses++;
      console.log(`✗ Cache MISS: ${cacheKey}`);

      const result = await this.db.query(
        'SELECT id, name, email, created_at, last_login FROM users WHERE id = $1',
        [userId]
      );

      if (result.rows.length === 0) {
        // Cache negative result
        await this.cache.setEx(cacheKey, 300, 'NOT_FOUND');
        return null;
      }

      const user = result.rows[0];

      // 3. Store in cache
      await this.cache.setEx(cacheKey, this.defaultTTL, JSON.stringify(user));
      console.log(`→ Cached: ${cacheKey} (TTL: ${this.defaultTTL}s)`);

      return user;

    } catch (error) {
      console.error('Error in getUser:', error);
      throw error;
    }
  }

  /**
   * Update user and invalidate cache
   */
  async updateUser(userId, name, email) {
    try {
      // 1. Update database (source of truth)
      await this.db.query(
        'UPDATE users SET name = $1, email = $2, updated_at = NOW() WHERE id = $3',
        [name, email, userId]
      );

      this.stats.writes++;

      // 2. Invalidate cache
      const cacheKey = `user:profile:${userId}`;
      const deleted = await this.cache.del(cacheKey);

      if (deleted > 0) {
        console.log(`⊗ Cache INVALIDATED: ${cacheKey}`);
      }

      return true;

    } catch (error) {
      console.error('Error in updateUser:', error);
      throw error;
    }
  }

  /**
   * Batch get users (optimized)
   */
  async getUsersBatch(userIds) {
    const keys = userIds.map(id => `user:profile:${id}`);

    // 1. Batch get from cache
    const cachedValues = await Promise.all(
      keys.map(key => this.cache.get(key))
    );

    const result = {};
    const missingIds = [];

    userIds.forEach((userId, index) => {
      const cached = cachedValues[index];
      
      if (cached && cached !== 'NOT_FOUND') {
        result[userId] = JSON.parse(cached);
        this.stats.hits++;
      } else {
        missingIds.push(userId);
        this.stats.misses++;
      }
    });

    // 2. Batch query database for cache misses
    if (missingIds.length > 0) {
      const dbResult = await this.db.query(
        'SELECT id, name, email, created_at FROM users WHERE id = ANY($1)',
        [missingIds]
      );

      // 3. Batch store to cache
      const cachePromises = [];
      
      for (const user of dbResult.rows) {
        const cacheKey = `user:profile:${user.id}`;
        result[user.id] = user;
        
        cachePromises.push(
          this.cache.setEx(cacheKey, this.defaultTTL, JSON.stringify(user))
        );
      }

      await Promise.all(cachePromises);
    }

    return result;
  }

  /**
   * Calculate cache hit rate
   */
  getHitRate() {
    const total = this.stats.hits + this.stats.misses;
    return total === 0 ? 0 : (this.stats.hits / total) * 100;
  }

  /**
   * Print cache statistics
   */
  printStats() {
    console.log('\n' + '='.repeat(50));
    console.log('Cache Statistics');
    console.log('='.repeat(50));
    console.log(`Hits:       ${this.stats.hits}`);
    console.log(`Misses:     ${this.stats.misses}`);
    console.log(`Writes:     ${this.stats.writes}`);
    console.log(`Hit Rate:   ${this.getHitRate().toFixed(2)}%`);
    console.log('='.repeat(50) + '\n');
  }
}

// Demo function
async function demo() {
  // Setup connections
  const redisClient = redis.createClient({
    url: 'redis://localhost:6379'
  });
  
  await redisClient.connect();

  const pgPool = new Pool({
    host: 'localhost',
    database: 'testdb',
    user: 'postgres',
    password: 'postgres',
    port: 5432
  });

  const repo = new CacheAsideRepository(redisClient, pgPool);

  console.log('\n🚀 Cache-Aside Pattern Demo (Node.js)\n');

  try {
    // Demo 1: First read - cache miss
    console.log('Demo 1: First read (cache miss)');
    console.log('-'.repeat(50));
    let user = await repo.getUser(1);
    console.log('User:', user);
    console.log();

    // Demo 2: Second read - cache hit
    console.log('Demo 2: Second read (cache hit)');
    console.log('-'.repeat(50));
    user = await repo.getUser(1);
    console.log('User:', user);
    console.log();

    // Demo 3: Third read - still cache hit
    console.log('Demo 3: Third read (still cache hit)');
    console.log('-'.repeat(50));
    user = await repo.getUser(1);
    console.log('User:', user);
    console.log();

    // Demo 4: Update - invalidate cache
    console.log('Demo 4: Update user (invalidates cache)');
    console.log('-'.repeat(50));
    await repo.updateUser(1, 'Updated Name', 'updated@example.com');
    console.log();

    // Demo 5: Read after update - cache miss
    console.log('Demo 5: Read after update (cache miss, fresh data)');
    console.log('-'.repeat(50));
    user = await repo.getUser(1);
    console.log('User:', user);
    console.log();

    // Demo 6: Batch operations
    console.log('Demo 6: Batch get users');
    console.log('-'.repeat(50));
    const users = await repo.getUsersBatch([1, 2, 3, 4, 5]);
    console.log(`Got ${Object.keys(users).length} users`);
    console.log();

    // Print statistics
    repo.printStats();

  } finally {
    await redisClient.quit();
    await pgPool.end();
  }
}

// Run demo
if (require.main === module) {
  demo().catch(console.error);
}

module.exports = { CacheAsideRepository };

File: patterns/cache-aside/examples/cache_aside.go
// Cache-Aside Pattern - Go Implementation
// ========================================

package main

import (
	"context"
	"encoding/json"
	"fmt"
	"time"

	"github.com/go-redis/redis/v8"
	"github.com/jackc/pgx/v4"
)

var ctx = context.Background()

// User represents user data
type User struct {
	ID        int       `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	CreatedAt time.Time `json:"created_at"`
}

// CacheAsideRepo implements cache-aside pattern
type CacheAsideRepo struct {
	cache      *redis.Client
	db         *pgx.Conn
	defaultTTL time.Duration
	stats      Stats
}

// Stats tracks cache performance
type Stats struct {
	Hits   int64
	Misses int64
	Writes int64
}

// NewCacheAsideRepo creates new repository
func NewCacheAsideRepo(redisClient *redis.Client, dbConn *pgx.Conn) *CacheAsideRepo {
	return &CacheAsideRepo{
		cache:      redisClient,
		db:         dbConn,
		defaultTTL: 1 * time.Hour,
	}
}

// GetUser retrieves user with cache-aside pattern
func (r *CacheAsideRepo) GetUser(userID int) (*User, error) {
	cacheKey := fmt.Sprintf("user:profile:%d", userID)

	// 1. Check cache first
	cached, err := r.cache.Get(ctx, cacheKey).Result()
	if err == nil {
		r.stats.Hits++
		fmt.Printf("✓ Cache HIT: %s\n", cacheKey)

		if cached == "NOT_FOUND" {
			return nil, nil
		}

		var user User
		if err := json.Unmarshal([]byte(cached), &user); err != nil {
			return nil, err
		}
		return &user, nil
	}

	// 2. Cache miss - query database
	r.stats.Misses++
	fmt.Printf("✗ Cache MISS: %s\n", cacheKey)

	var user User
	err = r.db.QueryRow(ctx, `
		SELECT id, name, email, created_at
		FROM users
		WHERE id = $1
	`, userID).Scan(&user.ID, &user.Name, &user.Email, &user.CreatedAt)

	if err == pgx.ErrNoRows {
		// Cache negative result
		r.cache.Set(ctx, cacheKey, "NOT_FOUND", 5*time.Minute)
		return nil, nil
	}
	if err != nil {
		return nil, err
	}

	// 3. Store in cache
	userData, _ := json.Marshal(user)
	r.cache.Set(ctx, cacheKey, userData, r.defaultTTL)
	fmt.Printf("→ Cached: %s (TTL: %v)\n", cacheKey, r.defaultTTL)

	return &user, nil
}

// UpdateUser updates user and invalidates cache
func (r *CacheAsideRepo) UpdateUser(userID int, name, email string) error {
	// 1. Update database
	_, err := r.db.Exec(ctx, `
		UPDATE users
		SET name = $1, email = $2, updated_at = NOW()
		WHERE id = $3
	`, name, email, userID)

	if err != nil {
		return err
	}

	r.stats.Writes++

	// 2. Invalidate cache
	cacheKey := fmt.Sprintf("user:profile:%d", userID)
	deleted, _ := r.cache.Del(ctx, cacheKey).Result()
	if deleted > 0 {
		fmt.Printf("⊗ Cache INVALIDATED: %s\n", cacheKey)
	}

	return nil
}

// GetHitRate calculates cache hit rate
func (r *CacheAsideRepo) GetHitRate() float64 {
	total := r.stats.Hits + r.stats.Misses
	if total == 0 {
		return 0
	}
	return float64(r.stats.Hits) / float64(total) * 100
}

// PrintStats prints cache statistics
func (r *CacheAsideRepo) PrintStats() {
	fmt.Println("\n" + "="*50)
	fmt.Println("Cache Statistics")
	fmt.Println("="*50)
	fmt.Printf("Hits:       %d\n", r.stats.Hits)
	fmt.Printf("Misses:     %d\n", r.stats.Misses)
	fmt.Printf("Writes:     %d\n", r.stats.Writes)
	fmt.Printf("Hit Rate:   %.2f%%\n", r.GetHitRate())
	fmt.Println("="*50)
}

func main() {
	// Setup Redis connection
	rdb := redis.NewClient(&redis.Options{
		Addr: "localhost:6379",
		DB:   0,
	})
	defer rdb.Close()

	// Setup PostgreSQL connection
	conn, err := pgx.Connect(ctx, "postgres://postgres:postgres@localhost:5432/testdb")
	if err != nil {
		panic(err)
	}
	defer conn.Close(ctx)

	repo := NewCacheAsideRepo(rdb, conn)

	fmt.Println("\n🚀 Cache-Aside Pattern Demo (Go)\n")

	// Demo 1: First read - cache miss
	fmt.Println("Demo 1: First read (cache miss)")
	fmt.Println("-" + "-"*49)
	user, _ := repo.GetUser(1)
	fmt.Printf("User: %+v\n\n", user)

	// Demo 2: Second read - cache hit
	fmt.Println("Demo 2: Second read (cache hit)")
	fmt.Println("-" + "-"*49)
	user, _ = repo.GetUser(1)
	fmt.Printf("User: %+v\n\n", user)

	// Demo 3: Update - invalidate cache
	fmt.Println("Demo 3: Update user (invalidates cache)")
	fmt.Println("-" + "-"*49)
	repo.UpdateUser(1, "Updated Name", "updated@example.com")
	fmt.Println()

	// Demo 4: Read after update - cache miss
	fmt.Println("Demo 4: Read after update (cache miss, fresh data)")
	fmt.Println("-" + "-"*49)
	user, _ = repo.GetUser(1)
	fmt.Printf("User: %+v\n\n", user)

	repo.PrintStats()
}
